lib/std/crypto/25519/edwards25519.zig

```diff
@@ -250,8 +252,150 @@ pub const Edwards25519 = struct {
+        const y = xmontToYmont(x) catch unreachable;
+        return .{ .x = x, .y = y, .not_square = not_square };
+    }
+
+    /// Map a 64-bit hash into an Edwards25519 point
+    pub fn fromHash(h: [64]u8) Edwards25519 {
+        const fe_f = Fe.fromBytes64(h);
+        var elr = elligator2(fe_f);
+
+        const y_sign = elr.not_square;
+        const y_neg = elr.y.neg();
+        elr.y.cMov(y_neg, @boolToInt(elr.y.isNegative()) ^ @boolToInt(y_sign));
+        return montToEd(elr.x, elr.y).clearCofactor();
+    }
+
+    fn stringToPoints(comptime n: usize, ctx: []const u8, s: []const u8) [n]Edwards25519 {
+        debug.assert(n <= 2);
+        const H = std.crypto.hash.sha2.Sha512;
+        const h_l: usize = 48;
+        var xctx = ctx;
+        var hctx: [H.digest_length]u8 = undefined;
+        if (ctx.len > 0xff) {
+            var st = H.init(.{});
+            st.update("H2C-OVERSIZE-DST-");
+            st.update(ctx);
+            st.final(&hctx);
+            xctx = hctx[0..];
+        }
+        const empty_block = [_]u8{0} ** H.block_length;
+        var t = [3]u8{ 0, n * h_l, 0 };
+        var xctx_len_u8 = [1]u8{@intCast(u8, xctx.len)};
+        var st = H.init(.{});
+        st.update(empty_block[0..]);
+        st.update(s);
+        st.update(t[0..]);
+        st.update(xctx);
+        st.update(xctx_len_u8[0..]);
+        var u_0: [H.digest_length]u8 = undefined;
+        st.final(&u_0);
+        var u: [n * H.digest_length]u8 = undefined;
+        var i: usize = 0;
+        while (i < n * H.digest_length) : (i += H.digest_length) {
+            mem.copy(u8, u[i..][0..H.digest_length], u_0[0..]);
+            var j: usize = 0;
+            while (i > 0 and j < H.digest_length) : (j += 1) {
+                u[i + j] ^= u[i + j - H.digest_length];
+            }
+            t[2] += 1;
+            st = H.init(.{});
+            st.update(u[i..][0..H.digest_length]);
+            st.update(t[2..3]);
+            st.update(xctx);
+            st.update(xctx_len_u8[0..]);
+            st.final(u[i..][0..H.digest_length]);
+        }
+        var px: [n]Edwards25519 = undefined;
+        i = 0;
+        while (i < n) : (i += 1) {
+            mem.set(u8, u_0[0 .. H.digest_length - h_l], 0);
+            mem.copy(u8, u_0[H.digest_length - h_l ..][0..h_l], u[i * h_l ..][0..h_l]);
+            px[i] = fromHash(u_0);
+        }
+        return px;
+    }
+
+    /// Hash a context `ctx` and a string `s` into an Edwards25519 point
+    ///
+    /// This function implements the edwards25519_XMD:SHA-512_ELL2_RO_ and edwards25519_XMD:SHA-512_ELL2_NU_
+    /// methods from the "Hashing to Elliptic Curves" standard document.
+    ///
+    /// Although not strictly required by the standard, it is recommended to avoid NUL characters in
+    /// the context in order to be compatible with other implementations.
+    pub fn fromString(comptime random_oracle: bool, ctx: []const u8, s: []const u8) Edwards25519 {
```

Comment:
Why `String`?
Maybe `fromBytes`? And `bytesToPoints`.

Reply:
`fromBytes()` already exists, and recovers a point from a fixed-length, packed representation.

"byte string" is the terminology used in the standard document. In protocols, these are constructed like strings, being of variable length and including delimiters. `bytesToPoint()` wouldn't work with quotient groups (e.g. ristretto) for which this operation is also defined.

Reply:
For an example protocol, see SPAKE2. `s` is literally set to the strings `M for SPAKE2` and `N for SPAKE2`, the purpose of the hash-to-curve operation being to generate a "nothing up my sleeve" pair of group elements for which the discrete log is not known.